core/keychain/mnemonic.go

```diff
@@ -0,0 +1,86 @@
+package keychain
+
+import (
+	"errors"
+
+	"github.com/tyler-smith/go-bip39"
+)
+
+type generateKeyFromMnemonicOpts struct {
+	override bool
+	mnemonic string
+	password string
+}
+
+var defaultMnemonicOpts = generateKeyFromMnemonicOpts{
+	override: false,
+	mnemonic: "",
+	password: "",
+}
+
+type GenerateKeyFromMnemonicOpts func(o *generateKeyFromMnemonicOpts)
+
+func WithMnemonic(mnemonic string) GenerateKeyFromMnemonicOpts {
+	return func(o *generateKeyFromMnemonicOpts) {
+		if mnemonic != "" {
+			o.mnemonic = mnemonic
+		}
+	}
+}
+
+func WithPassword(password string) GenerateKeyFromMnemonicOpts {
+	return func(o *generateKeyFromMnemonicOpts) {
+		if password != "" {
+			o.password = password
+		}
+	}
+}
+
+func WithOverride() GenerateKeyFromMnemonicOpts {
+	return func(o *generateKeyFromMnemonicOpts) {
+		o.override = true
+	}
+}
+
+// Generates a public/private key pair using ed25519 algorithm.
+// It stores it in the local db and returns the mnemonic.
+// If Mnemonic is a blank string, it generates a random one.
+// If there's already a key pair stored, it overrides it if override is set to true. Returns an error otherwise
+func (kc *keychain) GenerateKeyFromMnemonic(opts ...GenerateKeyFromMnemonicOpts) (string, error) {
+	o := defaultMnemonicOpts
+	for _, opt := range opts {
+		opt(&o)
+	}
+	if val, _ := kc.store.Get([]byte(PublicKeyStoreKey)); val != nil && o.override == false {
+		newErr := errors.New("Error while executing GenerateKeyFromMnemonic. Key pair already exists.")
+		return "", newErr
+	}
+
+	mnemonic := o.mnemonic
+
+	if mnemonic == "" {
+		entropy, err := bip39.NewEntropy(256)
+		if err != nil {
+			return "", err
+		}
+
+		mnemonic, err = bip39.NewMnemonic(entropy)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	seed, err := bip39.NewSeedWithErrorChecking(mnemonic, o.password)
+	if err != nil {
+		return "", err
+	}
+
+	compressedSeed := seed[:32]
```

Comment:
Added a comment in 4944baf
Did this because ed25519 expects the seed to be of size = 32bytes, but the bip39 implementation is fixed to return size=64.
Refactored this so that instead of taking a slice of the seed, we rehash it again but with the desired size.